keps/sig-api-machinery/1027-api-unions/README.md

```diff
@@ -193,123 +429,651 @@ each list item is made of:
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
 
-Here are the major milestones for this KEP:
-- Initial discussion happened a year before the creation of this kep:
-  https://docs.google.com/document/d/1lrV-P25ZTWukixE9ZWyvchfFR0NE2eCHlObiCUgNQGQ/edit#heading=h.w5eqnf1f76x5
-- Points made in the initial document have been improved and put into this kep,
-  which has approved by sig-api-machinery tech-leads
-- KEP has been implemented:
-  - logic mostly lives in sigs.k8s.io/structured-merge-diff
-  - conversion between schema and openapi definition are in k8s.io/kube-openapi
-  - core types have been modified in k8s.io/kubernetes
-- Feature is ready to be released in Beta in kubernetes 1.15
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: APIUnions
+  - Components depending on the feature gate: kube-apiserver
+
+Request handlers in the api server will call into union validation and
+normalization function from the structured-merge-diff repo when feature is
+enabled.
+
+
+###### Does enabling the feature change any default behavior?
+
+No, since all built in unions are currently validated in other ways.
+
+<!--
+Any change of default behavior may be surprising to users or break existing
+automations, so be extremely careful here.
+-->
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+Yes, requests will simply skip union validation and normalization.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Custom resources that were skipping union validation when when the feature was
+rolled back may have allowed invalid data to persist. These must be updated to
```

Comment:
I'm possibly ok with this, but I'm interested in whether @liggitt wants to hardcode a version of ratcheting validation for unions to ease this transition.  I think such a thing could help schema authors.

Reply:
I think union validation should be included in the list of types of schema changes we want to handle ratcheting validation for (https://github.com/kubernetes/kubernetes/issues/94060), but I wouldn't hard-code a special version of ratcheting for this feature.

Reply:
I'm actually more interested in the "I update a CRD to add union validation and have to deal with existing data" user story than the "I disable and then re-enable the feature" story. Does this design cover what is required to do *that* successfully?

Reply:
We don't have to tag existing unions with non-standard validation. So there's no reason to need a ratchet? We can't change the set of objects that are accepted or rejected, but I think it should be OK for the error messages to change wording.

Reply:
Don't CRDs have that problem in spades? There are loads of incompatible changes you can make that we don't stop you from making. I don't think we should have to solve that problem for unions.

Reply:
I've added a section on ratcheting validation to try to spell out that we are not commiting to ratcheting validation (and not implementing it for alpha), but should make it reasonably possible to do in the future if necessary. 

Let me know if this is an okay compromise or can be articulated better/differently